packages/hardhat/contracts/Distributor.sol

```diff
@@ -26,49 +26,92 @@
 */
 pragma solidity >=0.8.0 <0.9.0;
 
-interface IERC20 {
-  function transfer(address to, uint256 value) external returns (bool);
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 
-  function transferFrom(
-    address from,
-    address to,
-    uint256 value
-  ) external returns (bool);
-}
 /* 
 Implementation Resources: 
   https://github.com/banteg/disperse-research
 */
 contract Distributor {
-  event Distributed(address indexed sender, string message);
+    using SafeERC20 for IERC20;
 
-  function distributeEther(address[] memory recipients, uint256[] memory values)
-    external
-    payable
-  {
-    for (uint256 i = 0; i < recipients.length; i++) {
-      payable(recipients[i]).transfer(values[i]);
-    }
-    uint256 balance = address(this).balance;
-    if (balance > 0) {
-      payable(msg.sender).transfer(balance);
-    }
-    emit Distributed(msg.sender, "Distributed");
-  }
+    event ethDistributed(address indexed sender, string indexed name);
+    event tokenDistributed(
+        address indexed sender,
+        address indexed token,
+        string indexed name
+    );
+
+    function distributeEther(
+        address[] memory recipients,
+        uint256[] memory values,
+        string memory name
+    ) external payable {
+        require(
+            recipients.length == values.length,
+            "Invalid recipients to values ratio"
+        );
+
+        uint256 total = 0;
+        for (uint256 i = 0; i < recipients.length; i++) {
+            require(
+                recipients[i] != address(0),
+                "You can't distribute to the genesis address"
+            );
+            require(values[i] > 0, "Why waste gas for zero Ether send");
+            total += values[i];
+        }
+
+        require(msg.value >= total, "Not enough Ether to distribute");
```

Comment:
Could just be a hardhat node thing, but this requirement breaks the current tests in the latest tests on the dev branch